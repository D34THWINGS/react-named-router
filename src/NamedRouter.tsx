import React, { createContext, useContext } from 'react';
import { BrowserRouter, BrowserRouterProps } from 'react-router-dom';
import { Route } from 'react-router';
/* eslint-disable import/no-extraneous-dependencies */
import { RouteConfig } from 'react-router-config';
/* eslint-enable */

import { BaseRoutingContext, buildRoutingContext, RoutingContext } from './utils';

export interface NamedRouteConfig extends RouteConfig {
  name?: string;
  title?: string;
  routes?: NamedRouteConfig[];
}

export interface NamedRouterProps<TRouterProps = BrowserRouterProps> {
  routes: NamedRouteConfig[];
  routerComponent?: React.ComponentType<TRouterProps>;
  routerProps?: TRouterProps;
  children?: React.ReactNode;
}

export const routerContext = createContext<BaseRoutingContext | RoutingContext>(buildRoutingContext([]));
export const useNamedRouting = () => useContext(routerContext);

const NamedRouter = ({
  routerComponent: RouterComponent = BrowserRouter,
  routerProps,
  routes,
  children,
}: NamedRouterProps) => (
  <RouterComponent {...routerProps}>
    <Route>
      {({ history }) => (
        <routerContext.Provider value={buildRoutingContext(routes, history)}>
          {children}
        </routerContext.Provider>
      )}
    </Route>
  </RouterComponent>
);

export default NamedRouter;
