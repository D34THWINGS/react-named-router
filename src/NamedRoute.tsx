import React from 'react';
import { Route, RouteComponentProps, RouteProps } from 'react-router-dom';
import invariant from 'tiny-invariant';

import { NamedRouteConfig, useNamedRouting } from './NamedRouter';

export interface NamedRouteProps<TParams = object> extends Omit<RouteProps, 'path'> {
  name: string;
  children?: React.ReactNode;
}

export interface NamedRouteComponentProps<P, C, S> extends RouteComponentProps<P, C, S> {
  route: NamedRouteConfig;
}

export const NamedRoute: React.FC<NamedRouteProps> = ({
  name,
  children,
  ...otherProps
}) => {
  const context = useNamedRouting();

  invariant(context, 'You should not use <NamedRoute> outside a <NamedRouter>');

  const route = context.getRoute(name);

  return (
    <Route {...route} {...otherProps}>
      {children}
    </Route>
  );
};

NamedRoute.displayName = 'NamedRoute';
