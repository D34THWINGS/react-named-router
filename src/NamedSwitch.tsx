import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Location } from 'history';
import { matchPath, match as Match } from 'react-router';
import invariant from 'tiny-invariant';
import { useNamedRouting } from './NamedRouter';

export interface NamedSwitchProps {
  location?: Location;
  children?: React.ReactNode;
}

const NamedSwitch = ({ location: locationFromProps, children }: NamedSwitchProps) => {
  const context = useNamedRouting();

  invariant(context, 'You should not use <NamedSwitch> outside a <NamedRouter>');

  const location = locationFromProps || context.location;
  let element = null;
  let match: null | Match = null;

  React.Children.forEach(children, (child) => {
    if (match === null && React.isValidElement(child)) {
      element = child;

      const path = child.props.name ? context.getRawPath(child.props.name) : child.props.from;
      const route = child.props.name ? context.getRoute(child.props.name) : {};

      match = matchPath(location.pathname, { ...route, ...child.props, path });
    }
  });

  return match && element
    ? React.cloneElement(element, { location, computedMatch: match })
    : null;
};

export default NamedSwitch;
