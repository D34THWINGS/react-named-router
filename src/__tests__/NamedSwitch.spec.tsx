import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { StaticRouter, StaticRouterProps } from 'react-router';
import { NamedSwitch } from '../NamedSwitch';
import { NamedRouter, NamedRouteConfig } from '../NamedRouter';
import { NamedRoute } from '../NamedRoute';

Enzyme.configure({ adapter: new Adapter() });

describe('NamedSwitch', () => {
  it('should render only first matching route', () => {
    // Given
    const routerProps: StaticRouterProps = { location: '/foo/bar' };

    // When
    const wrapper = mount((
      <NamedRouter
        routes={[{ name: 'foo', path: '/foo' }, { name: 'bar', path: '/foo/bar' }]}
        routerComponent={StaticRouter}
        routerProps={routerProps}
      >
        <NamedSwitch>
          <NamedRoute name="foo" exact />
          <NamedRoute name="bar" />
        </NamedSwitch>
      </NamedRouter>
    ));

    // Then
    expect(wrapper.find('NamedSwitch')).toMatchSnapshot();
  });

  it('should spread matching route properties', () => {
    // Given
    const FooComponent: React.FC = () => <div>Foo</div>;
    const BarComponent: React.FC = () => <div>Bar</div>;
    const routerProps: StaticRouterProps = { location: '/foo/bar' };
    const routes: NamedRouteConfig[] = [
      { name: 'foo', path: '/foo', exact: true, component: FooComponent },
      { name: 'bar', path: '/foo/bar', component: BarComponent },
    ];

    // When
    const wrapper = mount((
      <NamedRouter
        routes={routes}
        routerComponent={StaticRouter}
        routerProps={routerProps}
      >
        <NamedSwitch>
          <NamedRoute name="foo" />
          <NamedRoute name="bar" />
        </NamedSwitch>
      </NamedRouter>
    ));

    // Then
    expect(wrapper.find('NamedSwitch')).toMatchSnapshot();
  });
});
