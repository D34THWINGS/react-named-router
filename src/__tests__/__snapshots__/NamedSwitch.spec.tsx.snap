// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamedSwitch should render only first matching route 1`] = `
<NamedRouter
  routerComponent={[Function]}
  routerProps={
    Object {
      "location": "/foo/bar",
    }
  }
  routes={
    Array [
      Object {
        "name": "foo",
        "path": "/foo",
      },
      Object {
        "name": "bar",
        "path": "/foo/bar",
      },
    ]
  }
>
  <StaticRouter
    location="/foo/bar"
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/foo/bar",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      staticContext={Object {}}
    >
      <Route>
        <NamedSwitch>
          <NamedRoute
            computedMatch={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/foo/bar",
                "url": "/foo/bar",
              }
            }
            location={
              Object {
                "hash": "",
                "pathname": "/foo/bar",
                "search": "",
                "state": undefined,
              }
            }
            name="bar"
          >
            <Route
              computedMatch={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/foo/bar",
                  "url": "/foo/bar",
                }
              }
              location={
                Object {
                  "hash": "",
                  "pathname": "/foo/bar",
                  "search": "",
                  "state": undefined,
                }
              }
              path="/foo/bar"
              route={
                Object {
                  "name": "bar",
                  "parents": Array [],
                  "path": "/foo/bar",
                  "regex": /\\^\\\\/foo\\\\/bar\\\\/\\?\\$/,
                }
              }
            />
          </NamedRoute>
        </NamedSwitch>
      </Route>
    </Router>
  </StaticRouter>
</NamedRouter>
`;
